### ISSUE TITLE
sql: Joining subqueries causes OOM

### ISSUE THREAD

**Issue by bdarnell** [cockroachdb/cockroach#21165]
A query of the form `SELECT * FROM (SELECT * FROM table1 WHERE...) JOIN (SELECT * FROM table2 WHERE...) USING (table2_primary_key_columns) LIMIT 1000` causes nodes to run out of memory. (We have more schema details and the exact query in a non-public bug report) It looks like something memory is not being accounted for correctly. (blaming this on the use of a join on two subqueries is just a guess on my part)

**Comment by petermattis** [comment]
@knz Can you triage? Why wasn't memory monitoring canceling this query?

**Comment by knz** [comment]
This likely a case of some distsql processor not performing memory accounting, or not completely.

**Comment by knz** [comment]
Yes so analysis of the issue reveals distsql is indeed enabled. In general we're not testing memory accounting on distsql-executed queries as thoroughly as we are testing locally-run queries.
To solve this particular problem we need to know which processor has caused the crash and equip it accordingly. (Alternatively, audit all the processors).

**Comment by knz** [comment]
Also as far as I can see the data we buffer in the now-unbounded/dynamic flow buffers is not accounted for. cc @RaduBerinde

**Comment by knz** [comment]
(So for clarity I do not think the problem is related to the use of subqueries. )

**Comment by benesch** [comment]
This query is also OOM'ing on register:

```
> explain (verbose) select count(*) from schemas;
+----------------+-------+--------+-------------+-----------------+---------+----------+
|      Tree      | Level |  Type  |    Field    |   Description   | Columns | Ordering |
+----------------+-------+--------+-------------+-----------------+---------+----------+
| group          |     0 | group  |             |                 | (count) |          |
|  │             |     0 |        | aggregate 0 | count_rows()    |         |          |
|  └── render    |     1 | render |             |                 | (count) |          |
|       └── scan |     2 | scan   |             |                 | (count) |          |
|                |     2 |        | table       | schemas@primary |         |          |
|                |     2 |        | spans       | ALL             |         |          |
+----------------+-------+--------+-------------+-----------------+---------+----------+
(6 rows)
```

**Comment by petermattis** [comment]
@benesch On the surface this doesn't seem related. The table scan should pipe directly into an aggregation. I'm very surprised we're seeing an OOM from a `select count(*)` query. Is there something here I'm missing?

**Comment by benesch** [comment]
Nope. Perhaps I misinterpreted @knz on Slack; I thought he thought they were related. FWIW @RaduBerinde is also very surprised that we're OOMing on COUNT(*).

**Comment by adg** [comment]
We were able to trigger an OOM of the cockroach server when performing a `DELETE FROM` with a subquery:

```
DELETE FROM blobs WHERE repo IN (SELECT repo FROM unlicensed);
```

Table structure:

```
> SHOW COLUMNS FROM blobs;
+-------+--------+------+---------+----------------------------------+
| Field |  Type  | Null | Default |             Indices              |
+-------+--------+------+---------+----------------------------------+
| repo  | STRING | true | NULL    | {"blobs_repo_hash_key"}          |
| hash  | STRING | true | NULL    | {"blobs_repo_hash_key","hashes"} |
+-------+--------+------+---------+----------------------------------+
> SHOW COLUMNS FROM unlicensed;
+-------+--------+-------+---------+-------------+
| Field |  Type  | Null  | Default |   Indices   |
+-------+--------+-------+---------+-------------+
| repo  | STRING | false | NULL    | {"primary"} |
+-------+--------+-------+---------+-------------+
```

Table `blobs` had about 100m rows. Table `unlicensed` had about 32k rows.

The explain of the query:

```
> EXPLAIN DELETE FROM blobs WHERE repo IN (SELECT repo FROM unlicensed);
+-------+--------+------------+---------------------------+
| Level |  Type  |   Field    |        Description        |
+-------+--------+------------+---------------------------+
|     0 | delete |            |                           |
|     0 |        | from       | blobs                     |
|     1 | render |            |                           |
|     2 | scan   |            |                           |
|     2 |        | table      | blobs@blobs_repo_hash_key |
|     2 |        | spans      | ALL                       |
|     2 |        | subqueries |                         1 |
|     3 | scan   |            |                           |
|     3 |        | table      | unlicensed@primary        |
|     3 |        | spans      | ALL                       |
+-------+--------+------------+---------------------------+
```

I was able to mitigate this by adding a `LIMIT 10000` to the query.

This may be a separate issue with a separate cause, as I assume the `DELETE FROM` happens inside an implicit transaction.

Note that there's no concurrent activity on these tables while the query is in progress.

**Comment by knz** [comment]
@solongordon  @arjunravinarayan can you help me triage this?

**Comment by petermattis** [comment]
@adg You're correct that the `DELETE FROM` will happen inside an implicit transaction. Cockroach 1.1.x has a limit on the size of transactions it can handle and unfortunately the checks that you don't exceed that limit are not always enforced. What version are you running? We've made progress in this area for the 2.0 release (about a month away). At the very least, we should be returning an error message rather then OOM'ing the server. 

Do you have an estimate of how many rows your `DELETE FROM` query was deleting? The `SELECT repo FROM unlicensed` query will return 32k rows, but how many rows do those repos match in the `blobs` table. 

Adding a `LIMIT` clause to the `DELETE` is a good workaround in the short term.

**Comment by dsymonds** [comment]
@adg and I are running CockroachDB 1.1.5.

I triggered the same OOM for a similar `DELETE FROM`, where there were about 2M rows to delete. I think @adg's `DELETE FROM` had something in the order of millions of rows to delete too.

It's likely tolerable for us for now, though we'll be looking forward to 2.0.

**Comment by petermattis** [comment]
@dsymonds Do you need to transactionally delete all of those rows?

**Comment by dsymonds** [comment]
We don't need this right now, no. Is there an easy way to have the same effect without transactions that we overlooked?

**Comment by petermattis** [comment]
There is no way to perform DML operations non-transactionally in SQL (at least, in Cockroach SQL), but if you don't need to delete all of the rows transactionally, a series of `DELETE FROM ... LIMIT x` statements outside of a transaction can workaround the transaction size limitation.

**Comment by adg** [comment]
@petermattis that's what we're doing now, however intuitively it seems to be less efficient in that it needs to re-start the table scan on each request.

**Comment by knz** [comment]
> intuitively it seems to be less efficient in that it needs to re-start the table scan on each request.

The overhead of restarting the scan is amortized by the cost of the writes for the deletion.

**Comment by knz** [comment]
(Assuming sufficiently large LIMIT: >2000 should be sufficient)

**Comment by dsymonds** [comment]
I've been finding these `DELETE FROM` statements to start stacking up. If I've killed my sql command line client, for instance, and run another `DELETE FROM` (with a limit anywhere from 1000 to 5000), I find the deletes get "stuck" and never complete. For instance, `SELECT COUNT(*) FROM [SHOW CLUSTER QUERIES]` shows a steadily-increasing number (it's now above 1200), despite me only having started (and abandoned from the client side) maybe only 5-10 queries. Is this a related problem? They're all in the "executing" phase (except for the `SELECT` itself). Is this some problem with the two phase commit somehow generating more and more statements?

**Comment by dsymonds** [comment]
(And doing a `CANCEL QUERY` on one or two of the oldest still-running queries, which have been "executing" for nearly 24h, seems to unblock and make all the others disappear from the `[SHOW CLUSTER QUERIES]` list. That seems to indicate there's some problem there. I've been running with `SET DEFAULT_TRANSACTION_ISOLATION TO SNAPSHOT;` to attempt to speed this up, but this stacking-up effect was happening before that too.)

**Comment by petermattis** [comment]
@dsymonds This sounds similar to another bug reported by a customer this past week. It would be helpful if you could provide reproduction details? I think we have enough from above and the other report, but reproduced bugs are soon-to-be-fixed bugs and any additional info would be appreciated.

PS `SNAPSHOT` isolation provides little benefit over `SERIALIZABLE` for Cockroach and receives a lot less attention in our testing. I'm not surprised it isn't helping.

**Comment by dsymonds** [comment]
Good to know it might be a known bug. Is there a link to it somewhere?

What additional info can I provide? It's perhaps infeasible to reproduce completely, since it is about ~30 GB of database. My query is `DELETE FROM cayley.quads WHERE subject_hash IN (SELECT hash FROM cayley.nodes WHERE regexp_extract(value_string, '...') IN (SELECT repo FROM git.unlicensed)) AND predicate_hash = b'...' LIMIT 1000;` (a couple of strings changed to `...`). The total number of rows (replacing `DELETE FROM` with `SELECT COUNT(*) FROM`) is about 2M, and the number of rows in the subquery is about 265k.

**Comment by knz** [comment]
We had a bug with SHOW QUERIES where the queries would not be removed during server-side automatic retries: #22574. Fixed in upcoming 2.0.

**Comment by petermattis** [comment]
> Good to know it might be a known bug. Is there a link to it somewhere?

Not yet (only a private bug report). 

> What additional info can I provide? It's perhaps infeasible to reproduce completely, since it is about ~30 GB of database. My query is DELETE FROM cayley.quads WHERE subject_hash IN (SELECT hash FROM cayley.nodes WHERE regexp_extract(value_string, '...') IN (SELECT repo FROM git.unlicensed)) AND predicate_hash = b'...' LIMIT 1000; (a couple of strings changed to ...). The total number of rows (replacing DELETE FROM with SELECT COUNT(*) FROM) is about 2M, and the number of rows in the subquery is about 265k.

The `DELETE ... LIMIT 1000` will limit the size of the deletion transaction, but those nested queries look expensive. The `SELECT hash FROM cayley.nodes` query looks like it will need to perform a full table scan. You can prepend `EXPLAIN` to that query to see the query plan without executing the query. Cockroach 2.0 should handle the large transaction better and perhaps obviate the need for the `LIMIT 1000` (we should verify that). 

Are you expecting to perform these `DELETE` transactions frequently, or is this a one-off query? An good short term mitigation would be to figure out how to construct this query so that it can be performed incrementally. The `LIMIT 1000` achieves that in a very inefficient form. Unfortunately, I'm not seeing a way to restructure the query to achieve that.

**Comment by dsymonds** [comment]
These deletions were a one off. We gave up for now and just started recreating the tables from scratch. We will likely have subquery joins in the future, though.

**Comment by jordanlewis** [comment]
DistSQL doesn't support subqueries yet, so the hypothesis stated at the top isn't correct. Because of this, I'm going to close this issue. We're not going to backport fixes to memory monitoring to the local execution engine.

**Comment by RaduBerinde** [comment]
`SELECT * FROM (SELECT * FROM table1 WHERE...) JOIN (SELECT * FROM table2 WHERE...) USING (table2_primary_key_columns) LIMIT 1000`

^ this query has no subqueries.

**Comment by jordanlewis** [comment]
Gotcha. Thanks for catching that.

**Comment by arjunravinarayan** [comment]
We've greatly improved the memory handling of DistSQL, and also transitioned all local query execution to DistSQL by default. Closing this issue as the reports are now tied closely to the older 1.1 version of the database.
